theon/src/allocator.rs

```diff
@@ -15,19 +18,28 @@ pub(crate) trait Heap {
     fn len(&self) -> usize;
 }
 
-/// A SliceHeap is a heap created by destructuring
+/// A Block is a heap created by destructuring
 /// the elements of a mutable slice.
-pub(crate) struct SliceHeap {
+pub(crate) struct Block {
     heap: *mut u8,
     len: usize,
 }
 
-impl SliceHeap {
-    pub fn new(arena: &mut [u8]) -> SliceHeap {
-        SliceHeap { heap: arena.as_mut_ptr(), len: arena.len() }
+impl Block {
+    pub fn new(arena: &mut [u8]) -> Block {
```

Comment:
Without getting into the details of the actual allocator implementation one high-level comment I would make would be to avoid `&mut [u8]` as much as possible within the implementation. It's going to be very difficult to ensure all the properties of a borrowed slice within an allocator. It can seem convenient to use safe slices, but once you start trying to uphold all the guarantees necessary to use them it quickly becomes more of a burden than a convenience.

I would suggest replacing any instance of `&mut [u8]` or `&[u8]` with either `NonNull<[u8]>` (NonNull::slice_from_raw_parts) or `*mut [u8]` (core::ptr::slice_from_raw_parts[_mut])

Reply:
So internally I think this is largely true. The only time I create slices are when allocating fresh memory from one of the free lists; I'll take a look at using raw slices or `NonNull<[u8]>` in those places.

Block's `new` takes a slice as a signal that it's taking ownership of a block of memory, but internally destructures it into a pointer and length pair.

Reply:
> Block's `new` takes a slice as a signal that it's taking ownership of a block of memory, but internally destructures it into a pointer and length pair.

It looked like that might have been the intent. The problem is that a slice is borrowed not owned. The owned type would be a `Box<[u8]>` which could work there. But again it can be very easy to run into undefined behavior when trying to coerce between safe types and unsafe pointers especially when it's not what you would consider the user visible api boundary of your library. I think the rust std library is the best example of this and they liberally use a  [Unique<T>](https://github.com/rust-lang/rust/blob/4d570eea025a19564429eb52b34ec34e14659f55/library/core/src/ptr/unique.rs#L36) type to represent a raw pointer that is semantically owned (but will not have any automatic Drop invocation). I've seen lots of projects just copy the bits of Unique<T> that make sense for them and that may be something worth considering especially for a project that deals with raw ptrs as often as a hypervisor.

Reply:
Ok, I think I get it. `Block` was kind of already an analogue of `Unique`, and I got rid of all use of slices everywhere in this code, both in the allocator and in Block's `new` method. What do you think?